lib/models/chat_history.dart

```diff
@@ -264,12 +247,10 @@ Stream<DeltaEvent> _handleDocumentChange(
         if (message is! TwitchHypeTrainEventModel) {
           return message;
         }
-        return TwitchHypeTrainEndEventModel(
-            pinned: true,
-            messageId: "train${data['event']['id']}",
-            level: data['event']['level'],
-            total: data['event']['total'],
-            wasSuccessful: message.progress >= message.total ? true : false);
+        return TwitchHypeTrainEndEventModel.fromDocument(
```

Comment:
Hmm this is kind of clever, I didn't really expect a use case where `UpdateDeltaEvent` would return a _different_ message model although I guess it works. Did you consider using the same model and rendering differently based on a getter? Might be less code but not sure.

Otherwise, looks good. :)

Reply:
Well I also think it is kind of weird, but two visual items might be "too much"? Maybe something like updating the previous model to be hidden and then just append the end model could work? (Is deletion even possible?)

Reply:
An `UpdateDeltaEvent` replaces the current event though, so it'll reuse the same widget. It shouldn't create a second visual item.

Deletion isn't possible right now, but in principle we could write a `DeleteDeltaEvent` and handle it too. I'm not sure it's necessary though...